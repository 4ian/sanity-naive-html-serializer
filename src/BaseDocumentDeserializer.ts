import schema from 'part:@sanity/base/schema'
import Schema from '@sanity/schema'
import blockTools from '@sanity/block-tools'
import {
  customDeserializers,
  customBlockDeserializers,
} from './BaseSerializationConfig'
import { ObjectField, ObjectSchemaType, BlockSchemaType } from '@sanity/types'
import { Deserializer } from './types'

const noSchemaWarning = (obj: Element) =>
  `WARNING: Unfortunately the deserializer may have issues with this field or object: ${obj.className}.
  If it's a specific type, you may need to declare  at the top level, or write a custom deserializer.`

export default function generateRandomHash() {
  const alphabet = 'abcdefghijklmnopqrstuvxyz'
  const numbers = '012345679'
  let output = ''
  for (let i = 0; i < 9; i += 1) {
    output += `${alphabet[Math.ceil(Math.random() * (alphabet.length - 1))]}${
      numbers[Math.ceil(Math.random() * (numbers.length - 1))]
    }`
  }
  return output
}

const defaultSchema = Schema.compile({
  name: 'default',
  types: [
    {
      type: 'object',
      name: 'default',
      fields: [
        {
          name: 'block',
          type: 'array',
          of: [
            {
              type: 'block',
              marks: {
                annotations: [
                  {
                    name: 'link',
                    type: 'object',
                    title: 'External Link',
                    fields: [
                      {
                        name: 'url',
                        type: 'url',
                      },
                    ],
                  },
                  {
                    name: 'internalLink',
                    type: 'object',
                    title: 'Internal link',
                    fields: [
                      {
                        name: 'internalLink',
                        type: 'internalLink',
                        title: 'Internal Page',
                      },
                    ],
                  },
                ],
              },
            },
          ],
        },
      ],
    },
  ],
})

const blockContentType = defaultSchema
  .get('default')
  .fields.find((field: ObjectField) => field.name === 'block').type

export const deserializeDocument = (
  serializedDoc: string,
  deserializers: Record<string, any> = customDeserializers,
  blockDeserializers = customBlockDeserializers
) => {
  const metadata: Record<string, any> = {}
  const head = new DOMParser().parseFromString(serializedDoc, 'text/html').head

  // meta
  Array.from(head.children).forEach(metaTag => {
    const validTags = ['_id', '_rev', '_type']
    const metaName = metaTag.getAttribute('name')
    if (metaName && validTags.includes(metaName)) {
      metadata[metaName] = metaTag.getAttribute('content')
    }
  })

  const content: Record<string, any> = deserializeHTML(
    serializedDoc,
    schema.get(metadata._type),
    deserializers,
    blockDeserializers
  )

  return {
    ...metadata,
    ...content,
  }
}

const deserializeHTML = (
  html: string,
  target: ObjectSchemaType | BlockSchemaType,
  deserializers: Record<string, any>,
  blockDeserializers: Array<any>
): Record<string, any> | any[] => {
  //parent node is always div with classname of field -- get its children
  let HTMLnode = new DOMParser().parseFromString(html, 'text/html').body
    .children[0]
  if (!HTMLnode) {
    return {}
  }
  //@ts-ignore
  let output

  //catch embedded object as a field -- not necessarily foolproof
  if (
    HTMLnode.children.length === 1 &&
    target &&
    target.type &&
    HTMLnode.children[0].className === target.type.name
  ) {
    HTMLnode = HTMLnode.children[0]
  }

  //prioritize custom deserialization
  if (
    target &&
    target.type &&
    target.type.name &&
    Object.keys(deserializers.types).includes(target.type.name)
  ) {
    const deserialize = deserializers.types[target.type.name]
    output = deserialize(HTMLnode)
  }

  //begin recursively deserializing
  //every child node is either a field in an object
  //or a child object in an array
  else {
    const children = Array.from(HTMLnode.children)
    output = target && target.type && target.type.hasOwnProperty('of') ? [] : {}

    children.forEach(child => {
      let deserializedObj
      //prioritize custom deserialization
      if (Object.keys(deserializers.types).includes(child.className)) {
        const deserialize = deserializers.types[child.className]
        deserializedObj = deserialize(child)
      }

      //flat string, it's an unrich field
      else if (
        child.tagName.toLowerCase() === 'span' ||
        // @ts-ignore
        (child.dataset && child.dataset.type === 'string')
      ) {
        deserializedObj = child.innerHTML
      }

      //has specific class name, so it's either a field or obj
      else if (child.className) {
        let objType
        //wrapper is fieldname
        if (target && target.fields) {
          objType = target.fields.find(field => field.name === child.className)
        }

        // Find if the field is a type of the schema.
        if (!objType && schema.get(child.className)) {
          objType = schema.get(child.className)
        }

        // Find if the field is a field of an object
        //@ts-ignore
        if (!objType && target && target.type && target.type.fields) {
          //@ts-ignore
          objType = target.type.fields.find(
            (field: ObjectField) => field.name === child.className
          )
        }

        // Find if the field is a field of an array
        //@ts-ignore
        if (!objType && target && target.type && target.type.of) {
          //@ts-ignore
          objType = target.type.of.find(
            (field: ObjectField) => field.name === child.className
          )
        }

        if (!objType) {
          console.debug(noSchemaWarning(child))
          objType = blockContentType
        }

        try {
          deserializedObj = deserializeHTML(
            child.outerHTML,
            objType,
            deserializers,
            blockDeserializers
          )
        } catch (err) {
          console.debug(err)
          try {
            deserializedObj = deserializeHTML(
              child.outerHTML,
              schema.get('object'),
              deserializers,
              blockDeserializers
            )
          } catch (err) {
            console.debug(
              `Tried to deserialize block of type ${child.className} but failed! Received error ${err}`
            )
          }
        }

        // @ts-ignore
        if (Array.isArray(output) && deserializedObj) {
          deserializedObj._type = child.className
          // make sure that a key is rendered
          deserializedObj._key = child.id || generateRandomHash()
        }
      } else {
        deserializedObj = blockTools.htmlToBlocks(
          child.outerHTML,
          blockContentType,
          { rules: blockDeserializers }
        )[0]
        // make sure that a key is rendered
        deserializedObj._key = child.id || generateRandomHash()
      }

      // @ts-ignore
      if (Array.isArray(output)) {
        output.push(deserializedObj)
      } else {
        // @ts-ignore
        output[child.className] = deserializedObj
      }
    })
  }

  return output
}

export const BaseDocumentDeserializer: Deserializer = {
  deserializeDocument,
  deserializeHTML,
}
