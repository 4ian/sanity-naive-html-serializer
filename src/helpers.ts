import sanityClient from 'part:@sanity/base/client'
import { SanityDocument } from '@sanity/types'

const client = sanityClient.withConfig({ apiVersion: '2021-03-25' })

export const findLatestDraft = (documentId: string, ignoreI18n = true) => {
  //eliminates i18n versions
  const query = `*[_id match $id ${
    ignoreI18n ? ' && (_id in path("drafts.*") || _id in path("*"))' : ''
  }]`
  const params = { id: `*${documentId}` }
  return client
    .fetch(query, params)
    .then(
      (docs: SanityDocument[]) =>
        docs.find(doc => doc._id.includes('draft')) ?? docs[0]
    )
}
