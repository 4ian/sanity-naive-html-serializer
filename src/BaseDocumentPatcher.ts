import { findLatestDraft } from './helpers'
import sanityClient from 'part:@sanity/base/client'
import { Patcher } from './types'

//TODO: move all logic around fetching the right base document
//and merging to the right target document outside of this class

const client = sanityClient.withConfig({ apiVersion: '2021-03-25' })

const getUrl = (path: string) => {
  const { projectId, apiVersion } = client.config()
  const baseUrl = `https://${projectId}.api.sanity.io/v${apiVersion}`
  return baseUrl + path
}

const findDocumentAtRevision = async (documentId: string, rev: string) => {
  const { dataset } = client.config()
  let path = `/data/history/${dataset}/documents/${documentId}?revision=${rev}`
  let url = getUrl(path)
  let revisionDoc = await fetch(url, { credentials: 'include' })
    .then(req => req.json())
    .then(req => req.documents[0])
  /* endpoint will silently give you incorrect doc
   * if you don't request draft and the rev belongs to a draft, so check
   */
  if (revisionDoc._rev !== rev) {
    path = `/data/history/${dataset}/documents/drafts.${documentId}?revision=${rev}`
    url = getUrl(path)
    revisionDoc = await fetch(url, { credentials: 'include' })
      .then(req => req.json())
      .then(req => req.documents[0])
  }
  return revisionDoc
}

const fieldLevelPatch = async (
  translatedFields: Record<string, any>,
  documentId: string,
  localeId: string,
  baseLang: string = 'en'
) => {
  let doc: Record<string, any>
  if (translatedFields._rev) {
    doc = await findDocumentAtRevision(documentId, translatedFields._rev)
  } else {
    doc = await findLatestDraft(documentId)
  }
  const merged: Record<string, any> = {}

  for (let field in translatedFields) {
    if (field === '_rev') {
      continue
    }
    const translatedVal = translatedFields[field][baseLang]
    const origVal = doc[field][baseLang]

    let valToPatch
    if (typeof translatedVal === 'string') {
      valToPatch = translatedVal
    } else if (Array.isArray(translatedVal)) {
      valToPatch = reconcileArray(origVal ?? [], translatedVal)
    } else {
      valToPatch = reconcileObject(origVal ?? {}, translatedVal)
    }
    merged[`${field}.${localeId.replace('-', '_')}`] = valToPatch
  }

  client
    .patch(doc._id)
    .set(merged)
    .commit()
}

const documentLevelPatch = async (
  translatedFields: Record<string, any>,
  documentId: string,
  localeId: string
) => {
  let baseDoc: Record<string, any> = await findLatestDraft(documentId)
  const targetId = `i18n.${documentId}.${localeId}`
  const i18nDoc = await findLatestDraft(targetId, false)

  if (translatedFields._rev) {
    baseDoc = await findDocumentAtRevision(documentId, translatedFields._rev)
  } else if (i18nDoc) {
    /* check for existing i18n version. and, if there's no rev, use it as base */
    baseDoc = baseDoc ?? i18nDoc
  }

  const merged = reconcileObject(baseDoc, translatedFields)
  if (i18nDoc) {
    const cleanedMerge: Record<string, any> = {}
    Object.entries(merged).forEach(([key, value]) => {
      if (Object.keys(translatedFields).includes(key)) {
        cleanedMerge[key] = value
      }
    })
    client
      .transaction()
      //@ts-ignore
      .patch(i18nDoc._id, p => p.set(cleanedMerge))
      .commit()
  } else {
    merged._id = `drafts.${targetId}`
    merged._lang = localeId
    client.create(merged)
  }
}

const reconcileArray = (origArray: any[], translatedArray: any[]) => {
  //arrays of strings don't have keys, so just replace the array and return
  if (translatedArray && translatedArray.some(el => typeof el === 'string')) {
    return translatedArray
  }

  //deep copy needed for field level patching
  const combined = JSON.parse(JSON.stringify(origArray))

  translatedArray.forEach(block => {
    if (!block._key) {
      return
    }
    const foundBlockIdx = origArray.findIndex(
      origBlock => origBlock._key === block._key
    )
    if (foundBlockIdx < 0) {
      console.log(
        `This block no longer exists on the original document. Was it removed? ${JSON.stringify(
          block
        )}`
      )
    } else if (
      origArray[foundBlockIdx]._type === 'block' ||
      origArray[foundBlockIdx]._type === 'span'
    ) {
      combined[foundBlockIdx] = block
    } else if (Array.isArray(origArray[foundBlockIdx])) {
      combined[foundBlockIdx] = reconcileArray(origArray[foundBlockIdx], block)
    } else {
      combined[foundBlockIdx] = reconcileObject(origArray[foundBlockIdx], block)
    }
  })
  return combined
}

const reconcileObject = (
  origObject: Record<string, any>,
  translatedObject: Record<string, any>
) => {
  const updatedObj = JSON.parse(JSON.stringify(origObject))
  Object.entries(translatedObject).forEach(([key, value]) => {
    if (!value || key[0] === '_') {
      return
    }
    if (typeof value === 'string') {
      updatedObj[key] = value
    } else if (Array.isArray(value)) {
      updatedObj[key] = reconcileArray(origObject[key] ?? [], value)
    } else {
      updatedObj[key] = reconcileObject(origObject[key] ?? {}, value)
    }
  })
  return updatedObj
}

export const BaseDocumentPatcher: Patcher = {
  fieldLevelPatch,
  documentLevelPatch,
  reconcileArray,
  reconcileObject,
}
