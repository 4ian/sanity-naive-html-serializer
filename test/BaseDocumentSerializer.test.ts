import { BaseDocumentSerializer } from '../src/BaseDocumentSerializer'
import { defaultStopTypes, customSerializers } from '../src/BaseSerializationConfig'
const documentLevelArticle = require('./mocks/articles/documentLevelArticle')
const fieldLevelArticle = require('./mocks/articles/fieldLevelArticle')

test('Global test of working doc-level functionality and snapshot match', async () => {
  const serializer = BaseDocumentSerializer
  const serialized = await serializer.serializeDocument(
    documentLevelArticle._id, 'document', 'en', defaultStopTypes, customSerializers)
  expect(serialized).toMatchSnapshot()
})

test('Global test of working field-level functionality and snapshot match', async () => {
  const serializer = BaseDocumentSerializer
  const serialized = await serializer.serializeDocument(
    fieldLevelArticle._id, 'field', 'en', defaultStopTypes, customSerializers)
  expect(serialized).toMatchSnapshot()
})

//(maybe overkill, but maybe good unit testing and/or for further debugging):
//expect individual fields and nested fields to be rendered as they are in the snapshot
//expect values in a list to not be repeated (this was a bug)

//more integration-y tests below:
//expect custom serializers to work
//expect localize false fields to be absent
//expect explicitly declared stop types to be absent