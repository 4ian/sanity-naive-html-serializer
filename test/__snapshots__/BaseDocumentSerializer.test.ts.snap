// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global test of working doc-level functionality and snapshot match 1`] = `
Object {
  "content": "<div class=\\"documentLevelArticle\\" id=\\"tsj8si\\"><span class=\\"title\\">My Document-Level Article</span><span class=\\"snippet\\">This is text in my text field.</span><div class=\\"tags\\"><span>tag 1</span><span>tag 2</span><span>tag 3</span></div><span class=\\"hidden\\"></span><div class=\\"config\\" id=\\"6rybyk\\"><div class=\\"nestedArrayField\\"><p id=\\"4a58adc7c507\\">This is block text 1 level deep</p></div><div class=\\"objectAsField\\"><div class=\\"childObjectField\\" id=\\"mrqq84\\"><span class=\\"title\\">This is one level deeper</span><div class=\\"content\\"><p id=\\"1aa228ac848b\\">This is a block text 2 levels deep</p></div></div></div><span class=\\"title\\">This is an object nested in a document</span></div><div class=\\"content\\"><p id=\\"a4d177e92666\\">This is block text at the top level.</p><div class=\\"objectField\\" id=\\"c0313627775e\\"><span class=\\"title\\">This is an object in top-level block text.</span><div class=\\"objectAsField\\"><div class=\\"childObjectField\\" id=\\"88za05\\"><span class=\\"title\\">This is a nested object in an object in top-level block text.</span><div class=\\"content\\"><p id=\\"4f24f6fbfae7\\">This is block text in a nested object in an object in top-level block text.</p></div></div></div></div></div></div>",
  "name": "drafts.d8ffc675-ce86-4f60-9ac8-da164cde3b0a",
}
`;

exports[`Global test of working field-level functionality and snapshot match 1`] = `
Object {
  "content": "<div class=\\"fieldLevelArticle\\" id=\\"mx623e\\"><div class=\\"config\\" id=\\"49a835\\"><div class=\\"en\\"><div class=\\"objectField\\" id=\\"v1764a\\"><span class=\\"title\\">This is an object nested in a document</span><div class=\\"objectAsField\\"><div class=\\"childObjectField\\" id=\\"dme9mc\\"><span class=\\"title\\">This is one level deeper</span><div class=\\"content\\"><p id=\\"869168fe3a2a\\">This is block text 2 levels deep</p></div></div></div><div class=\\"nestedArrayField\\"><p id=\\"f49b4d7e3e51\\">This is block text 1 level deep</p></div></div></div></div><div class=\\"content\\" id=\\"dcjm38\\"><div class=\\"en\\"><p id=\\"e2a39d768ff8\\">This is block text at the top level.</p><div class=\\"objectField\\" id=\\"271b0c6ee984\\"><span class=\\"title\\">This is an object in top-level block text.</span><div class=\\"objectAsField\\"><div class=\\"childObjectField\\" id=\\"88za05\\"><span class=\\"title\\">This is a nested object in an object in top-level block text.</span><div class=\\"content\\"><p id=\\"4f24f6fbfae7\\">This is block text in a nested object in an object in top-level block text.</p></div></div></div></div></div></div><div class=\\"snippet\\" id=\\"yq23r4\\"><span class=\\"en\\">This is text in my text field</span></div><div class=\\"tags\\" id=\\"jkg4ql\\"><div class=\\"en\\"><span>tag 1</span><span>tag 2</span><span>tag 3</span></div></div><div class=\\"title\\" id=\\"ulpd14\\"><span class=\\"en\\">My Field-Level Article</span></div></div>",
  "name": "drafts.2947533e-1ea5-4116-955b-339608d3445d",
}
`;
